File: runtime/ts/log.ts
const __almd_log = {
  debug(msg: string): void { console.error(`[DEBUG] ${msg}`); },
  info(msg: string): void { console.error(`[INFO] ${msg}`); },
  warn(msg: string): void { console.error(`[WARN] ${msg}`); },
  error(msg: string): void { console.error(`[ERROR] ${msg}`); },
  debug_with(msg: string, fields: [string, string][]): void { const kv = fields.map(([k,v]) => `${k}=${v}`).join(" "); console.error(`[DEBUG] ${msg} ${kv}`); },
  info_with(msg: string, fields: [string, string][]): void { const kv = fields.map(([k,v]) => `${k}=${v}`).join(" "); console.error(`[INFO] ${msg} ${kv}`); },
  warn_with(msg: string, fields: [string, string][]): void { const kv = fields.map(([k,v]) => `${k}=${v}`).join(" "); console.error(`[WARN] ${msg} ${kv}`); },
  error_with(msg: string, fields: [string, string][]): void { const kv = fields.map(([k,v]) => `${k}=${v}`).join(" "); console.error(`[ERROR] ${msg} ${kv}`); },
};

File: runtime/rs/src/log.rs
// log extern — structured logging to stderr

pub fn almide_rt_log_debug(msg: String) { eprintln!("[DEBUG] {}", msg); }
pub fn almide_rt_log_info(msg: String) { eprintln!("[INFO] {}", msg); }
pub fn almide_rt_log_warn(msg: String) { eprintln!("[WARN] {}", msg); }
pub fn almide_rt_log_error(msg: String) { eprintln!("[ERROR] {}", msg); }

pub fn almide_rt_log_debug_with(msg: String, data: String) { eprintln!("[DEBUG] {} {}", msg, data); }
pub fn almide_rt_log_info_with(msg: String, data: String) { eprintln!("[INFO] {} {}", msg, data); }
pub fn almide_rt_log_warn_with(msg: String, data: String) { eprintln!("[WARN] {} {}", msg, data); }
pub fn almide_rt_log_error_with(msg: String, data: String) { eprintln!("[ERROR] {} {}", msg, data); }
